### PR TITLE
[mlir][tosa] Refactor conversions to use templates

### PR DESCRIPTION
**Pull request by sjarus**
- Remove use of conversion construction macros
- Add mul and div op conversions
- Add corresponding tests

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

### CODE REVIEW

**Comment by sjarus** [comment]
@silvasean refactored as promised earlier. Added a couple of tests too.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -191,6 +192,71 @@ ConvertAtenReluOp::matchAndRewrite(AtenReluOp op, ArrayRef<Value> operands,
...
+  auto lhsElemTy = lhsTy.getElementType();
+  auto rhsElemTy = rhsTy.getElementType();
+
+  if (lhsElemTy != rhsElemTy)
+    return op.emitError("Add: input datatypes mismatched");
+
+  if (lhsElemTy.isa<mlir::FloatType>()) {
+    rewriter.replaceOpWithNewOp<tosa::MulOp>(
+        op, getTypeConverter()->convertType(op.getType()), lhs, rhs,
+        0 /*shift*/);
```
Comment (silvasean) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:218]: nit: in LLVM coding style, we usually do `/*shift=*/0`

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -191,6 +192,71 @@ ConvertAtenReluOp::matchAndRewrite(AtenReluOp op, ArrayRef<Value> operands,
...
+
+  if (lhsElemTy != rhsElemTy)
+    return op.emitError("Add: input datatypes mismatched");
+
+  if (lhsElemTy.isa<mlir::FloatType>()) {
+    auto rcpOp = rewriter.create<tosa::ReciprocalOp>(
+        op->getLoc(), getTypeConverter()->convertType(op.getType()), rhs);
+    rewriter.replaceOpWithNewOp<tosa::MulOp>(
+        op, getTypeConverter()->convertType(op.getType()), lhs,
+        rcpOp.getResult(), 0 /*shift*/);
```
Comment (silvasean) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:252]: same here

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -23,116 +23,112 @@ using namespace mlir::torch;
...
+    if (!lhsTy || !rhsTy)
+      return op.emitError("Only Tensor types supported in TOSA");
+
+    auto lhsElemTy = lhsTy.getElementType();
+    auto rhsElemTy = rhsTy.getElementType();
+
+    if (lhsElemTy != rhsElemTy)
+      return op.emitError("Add: input datatypes mismatched");
+
+    /* FIXME: Handle alpha. \
```
Comment (silvasean) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:100]: nit: use `//` comments.

File: test/Conversion/TorchToTosa/basic.mlir [hunk 4]
```diff
@@ -130,9 +130,40 @@ func @torch.aten.add$basic(%arg0: !torch.vtensor<[?, ?],f32>, %arg1: !torch.vten
...
+// CHECK:           %[[ARG0_BUILTIN:.*]] = torch_c.to_builtin_tensor %[[ARG0]] : !torch.vtensor<[?,?],f32> -> tensor<?x?xf32>
+// CHECK:           %[[ARG1_BUILTIN:.*]] = torch_c.to_builtin_tensor %[[ARG1]] : !torch.vtensor<[?,?],f32> -> tensor<?x?xf32>
+// CHECK:           %[[RCP:.*]] = "tosa.reciprocal"(%[[ARG1_BUILTIN]]) : (tensor<?x?xf32>) -> tensor<?x?xf32>
+// CHECK:           %[[RESULT_BUILTIN:.*]] = "tosa.mul"(%[[ARG0_BUILTIN]], %[[RCP]]) {shift = 0 : i32} : (tensor<?x?xf32>, tensor<?x?xf32>) -> tensor<?x?xf32>
+// CHECK:           %[[RESULT:.*]] = torch_c.from_builtin_tensor %[[RESULT_BUILTIN]] : tensor<?x?xf32> -> !torch.vtensor<[?,?],f32>
+// CHECK:           return %[[RESULT]] : !torch.vtensor<[?,?],f32>
+func @torch.aten.div$basic(%arg0: !torch.vtensor<[?, ?],f32>, %arg1: !torch.vtensor<[?, ?],f32>) -> !torch.vtensor<[?, ?],f32> {
+  %0 = torch.aten.div.Tensor %arg0, %arg1 : !torch.vtensor<[?, ?],f32>, !torch.vtensor<[?, ?],f32> -> !torch.vtensor<[?, ?],f32>
+  return %0 : !torch.vtensor<[?, ?],f32>
+}
```
Comment (silvasean) [inline test/Conversion/TorchToTosa/basic.mlir:169]: nit: newline at end of file (github is showing a scary red icon here)

**Review by silvasean — APPROVED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

**Comment by silvasean** [comment]
@sjarus FYI for the alpha handling, we have some helpers for this in TorchToLinalg that we might want to hoist somewhere common in the TorchConversion dialect. E.g. https://github.com/llvm/torch-mlir/blob/628a21bb1397470bcfa9ee8f2a24fac2ee12f28c/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1257

**Comment by sjarus** [comment]
> @sjarus FYI for the alpha handling, we have some helpers for this in TorchToLinalg that we might want to hoist somewhere common in the TorchConversion dialect. E.g.
> 
> https://github.com/llvm/torch-mlir/blob/628a21bb1397470bcfa9ee8f2a24fac2ee12f28c/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1257

That would be nice! I was looking for an m_TorchConstantFloat like there is for Int and Bool (and IntList). With something like this, it should be a simple update to the Add/Sub legalization common code.

**Comment by sjarus** [comment]
Related question - is there a nice way to combine target.addIllegalOp<>() and pattern.add<>() ? I'd get rid of the macros altogether if I could do that nicely - right now it's there just so I don't forget the addIllegalOp while adding a bunch of patterns.

**Comment by silvasean** [comment]
> Related question - is there a nice way to combine target.addIllegalOp<>() and pattern.add<>() ? I'd get rid of the macros altogether if I could do that nicely - right now it's there just so I don't forget the addIllegalOp while adding a bunch of patterns.

A lightweight variadic template is what I've used in the past. An example is setupFinalization here: https://github.com/llvm/torch-mlir/blob/628a21bb1397470bcfa9ee8f2a24fac2ee12f28c/lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp#L256

> That would be nice! I was looking for an m_TorchConstantFloat like there is for Int and Bool (and IntList). With something like this, it should be a simple update to the Add/Sub legalization common code.

Oh yes, I forgot that you are putting it in an attribute, so that function won't work for you. Indeed, adding m_TorchConstantFloat sounds like the right approach for TOSA right now. If possible, it would be nice if TOSA were somewhat less restrictive in this regard, though I think I understand the reason for the limitation.

**Comment by sjarus** [comment]
Well actually I'm not sure alpha needs to be either an attribute here really - I could just cast it as a scalar tensor and pass it to tosa.mul before the add/sub. 

On a related note however, I'm not sure about the datatype - the test in both LinAlg and Tosa assumes alpha can be an int even if self() is f32 . Should they match, or should TOSA attempt to cast from int to fp if they don't ?

**Comment by silvasean** [comment]
> On a related note however, I'm not sure about the datatype - the test in both LinAlg and Tosa assumes alpha can be an int even if self() is f32 . Should they match, or should TOSA attempt to cast from int to fp if they don't ?

It is allowed to be !torch.int and !torch.float, with the expectation that it is treated as a real value.

You can see in JITOperatorRegistryDump.txt how it is defined. `Scalar` mean `!torch.int` or `!torch.float` with expectation that it is treated as an approximate real number.
```
JitOperator 'aten::add.Tensor : (Tensor, Tensor, Scalar) -> (Tensor)':
  MLIR op name = torch.aten.add.Tensor
  MLIR td def name = Torch_AtenAddTensorOp
  namespace = aten
  unqualified_name = add
  overload_name = Tensor
  is_c10_op = True
  is_vararg = False
  is_varret = False
  is_mutable = False
  arguments:
    arg: {'name': 'self', 'type': 'Tensor', 'pytype': 'Tensor'}
    arg: {'name': 'other', 'type': 'Tensor', 'pytype': 'Tensor'}
    arg: {'name': 'alpha', 'type': 'Scalar', 'pytype': 'number', 'default_debug': '1'}
  returns:
    ret: {'name': '', 'type': 'Tensor', 'pytype': 'Tensor'}
```

**Comment by sjarus** [comment]
Thanks! So I'll just cast fo float for the current baseline FloatType handling.

**Comment by silvasean** [comment]
Hi Suraj, I noticed that after your recent PR's, ElementwiseAddModule_basic didn't start passing. Is this expected?

**Comment by sjarus** [comment]
That's a really good question - I actually added legalizations expecting to get XPASSes for everything, whereupon I allowlisted them. But in some cases I didn't see anything happen. This is probably one of them.

Looking at the test though, it seems a different signature - it just passes two parameters. Maybe it emits an aten.add.int ? I'm handling aten.add.Tensor.

**Comment by silvasean** [comment]
I think that one does go through aten.add.Tensor. Might be worth debugging more. If you pass `-v` to `./tools/torchscript_e2e_test.sh --config=tosa --filter=ElementwiseAddModule_basic` it should give a pretty readable description of how to reproduce the issue (or if not, let me know and I can make it better).

**Comment by sjarus** [comment]
Maybe I'm missing something here but I get just a summary output : 
$ ./tools/torchscript_e2e_test.sh -v --config=tosa --filter=ElementwiseAddModule_basic
XFAIL - "ElementwiseAddModule_basic"

Summary:
    Expectedly Failed: 1

Should I expect to see more than that ?

**Comment by ramiro050** [comment]
> Maybe I'm missing something here but I get just a summary output : $ ./tools/torchscript_e2e_test.sh -v --config=tosa --filter=ElementwiseAddModule_basic XFAIL - "ElementwiseAddModule_basic"
> 
> Summary: Expectedly Failed: 1
> 
> Should I expect to see more than that ?

The tests only print out debug information if the *unexpectedly* fail. If you add the corresponding module to your `TOSA_PASS_SET`, you should see the error message with instructions on how to reproduce it.

**Comment by sjarus** [comment]
Thanks @ramiro050 ! With that addition, I see that the failure is in TOSA to LinAlg-on-Tensors: 

        Exception:
        Lowering TOSA to Linalg-on-Tensors failed with the following diagnostics:
        error: 'linalg.generic' op expected indexing_map #1 to have 1 dim(s) to match the number of loops
        note: see current operation: %3 = "linalg.generic"(%arg0, %arg1, %2) ( {
        ^bb0(%arg2: f32, %arg3: f32, %arg4: f32):  // no predecessors
          %4 = "arith.addf"(%arg2, %arg3) : (f32, f32) -> f32
          "linalg.yield"(%4) : (f32) -> ()
        }) {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<() -> ()>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"], operand_segment_sizes = dense<[2, 1]> : vector<2xi32>} : (tensor<?xf32>, tensor<f32>, tensor<?xf32>) -> tensor<?xf32>

@silvasean @rsuderman Maybe this needs the TosaMakeBroadcastable pass to be called between Torch->TOSA and TOSA->LinAlg ?

**Comment by silvasean** [comment]
@sjarus Do you have the bandwidth to work on this? I suspect there are going to be multiple issues like this that require changing Tosa->LinAlg or other upstream TOSA passes, and it's not going to be very efficient to bounce bugs back and forth among each other.

I was hoping that ARM's contributions could take the form of vertical slices through all those passes, possibly needing to debug at various points in the stack, but the end result being new passing end-to-end tests at each commit, rather than simply isolated Torch->TOSA conversions.

Were you able to test and verify your lowering on the internal testing you said that you folks were bringing up?

**Comment by sjarus** [comment]
I'm not really familiar with the TosaToLinAlg pass so I'm hoping @rsuderman can help. 

If we can dump the TOSA form of this e2e test into a .mlir file we can decouple this investigation from the Torch-to-TOSA work. Is there a way to do that ?

**Comment by silvasean** [comment]
I'm generally really uncomfortable with having any lowering in tree that has not been tested end-to-end at least once to verify numerical correctness, so I really want to avoid decoupling these things (unless you have some internal end-to-end testing you're doing). It is really costly from a project perspective to have code in tree that has not received that kind of testing, as the risk of miscompile is extremely high -- this shakes out in integration as miscompiles that can take weeks to debug; it's far better to do the extra work up front to avoid these nightmare scenarios.

I'm ok with you using your internal testing system that you mentioned or `tools/torchscript_e2e_test.sh`.

I would really encourage you to work these as vertical slices, bringing online the relevant end-to-end tests in the same commit that adds the TorchToTosa lowering. The TosaToLinalg pass isn't very complicated (you might need to refresh on [anatomy of a linalg op](https://www.youtube.com/watch?v=PdQPlPudT90)).

**Comment by sjarus** [comment]
Yeah I'm completely in agreement here @silvasean . It's quite reasonable to turn off the AddTensor legalization (and any other similar ones) until it's established why it doesn't work further downstream in MLIR core. 

The functional verification at the TOSA level wouldn't catch this if this is in fact a LinAlg translation issue, but I'll push this independently after commenting out the legalizations that don't start working end-to-end.

**Comment by silvasean** [comment]
Cool, thanks! I'm not actually worried about TosaToLinalg working per se, but merely as a vehicle for some kind of end-to-end testing. As long as your functional verification has tested these TorchToTosa pattern, I'm okay keeping them enabled. I just want to make sure that *something* has run end-to-end through the pattern at least once (with correct results, of course) before committing in tree.

**Comment by sjarus** [comment]
It sounds like this topic could use a document describing the TOSA testing regime. For example, for the TF->TOSA path, nothing goes out into open source from our end without detailed bit-level validation of every legalization up to TOSA. It also lets us catch any legalization drift that happens outside. That's what we set up for Torch too, at least internally. 

I'll open an issue for this to document what we have. @armkevincheng is out today but we plan to coordinate the upstreaming of this into Torch-MLIR TOSA e2e CI next week. It won't address the LinAlg path as mentioned earlier but would explain the level of rigor for TOSA generation. 

I'll write up an issue on this and we can continue the conversation there. We're also keen that the Torch->TOSA->LinAlg path is stable (for our own needs), and any community input on how to ensure it doesn't break and how to fix breakages, would be useful.

**Comment by silvasean** [comment]
Thanks! That sounds perfect.

**Pull request merged**